lib/filterx/func-str.c

```diff
@@ -0,0 +1,387 @@
+/*
+ * Copyright (c) 2024 Axoflow
+ * Copyright (c) 2024 Szilard Parrag
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "filterx/func-str.h"
+#include "filterx/object-primitive.h"
+#include "filterx/filterx-object.h"
+#include "filterx/expr-literal.h"
+#include "filterx/expr-literal-generator.h"
+#include "filterx/filterx-eval.h"
+#include "object-extractor.h"
+#include "object-string.h"
+#include "filterx/object-json.h"
+#include "filterx/object-list-interface.h"
+
+#define FILTERX_FUNC_STARTSWITH_USAGE "Usage: startswith(string, prefix, ignorecase=true)" \
+"or startswith(string, [prefix_1, prefix_2, ..], ignorecase=true)"
+
+typedef struct _FilterXStringWithCache
+{
+  FilterXExpr *expr;
+  FilterXObject *obj;
+  const gchar *borrowed_str_value;
+  gchar *owned_str_value;
+  gsize str_len;
+} FilterXStringWithCache;
+
+typedef gboolean (*FilterXExprAffixProcessFunc)(const gchar *haystack, gsize haystack_len, const gchar *needle,
+                                                gsize needle_len);
+typedef struct _FilterXExprAffix
+{
+  FilterXFunction super;
+  gboolean ignore_case;
+  FilterXStringWithCache *haystack;
+  struct
+  {
+    FilterXExpr *expr;
+    GPtrArray *needle_objects_with_cache;
```

Comment:
maybe we could drop the needle prefix from the name of the array? we already have that in the enclosing struct.